octodns/provider/base.py

```diff
@@ -34,21 +34,29 @@ def __init__(self, id, apply_disabled=False,
 
     def _process_desired_zone(self, desired):
         '''
-        An opportunity for providers to modify that desired zone records before
-        planning.
-
-        - Must do their work and then call super with the results of that work
-        - Must not modify the `desired` parameter or its records and should
-          make a copy of anything it's modifying
+        An opportunity for providers to modify the desired zone records before
+        planning. `desired` is a "shallow" copy, see `Zone.copy` for more
+        information
+
+        - Must do their work and then call `super` with the results of that
+          work, returning the result of the `super` call.
+        - Must not modify `desired` directly, should call `desired.copy` and
+          modify the shallow copy returned from that.
```

Comment:
This is the main bit I've gone back and forth on. Part of me wants to call `_process_desired_zone` with a copy in the first place so that callers don't have to worry about it. The slightly funky part of that is things are then modifying a method parameter and we would probably want to get rid of the return value since it'd just be what was passed in. 

The pro would be that implementations don't have to make the copy and when the method is overridden there won't be multiple shallow copies made. The cons are that it would always be making a single copy and the aforementioned modifying a parameter thing.

Same things would be true for processors, and we should probably do the same there as here, but in that case there could be a lot of (shallow) copies made, one for each processor.

Reply:
Thought more about this and decided that it is best to create a single copy early on and pass it through everything after that allowing them to modify. As implmented in what's being pushed now that applies to both `Provider._process_source_zone` and the processors. Also added a bunch of documentation to all of these methods about the dos and don'ts.